crates/apollo-router-core/src/traits.rs

```diff
@@ -4,6 +4,12 @@ use futures::Future;
 use std::sync::Arc;
 use std::{fmt::Debug, pin::Pin};
 
+/// A planner key.
+///
+/// This type consists of a query string, an optional operation string and the
+/// [`QueryPlanOptions`].
+pub(crate) type QueryKey = (String, Option<String>, QueryPlanOptions);
```

Comment:
Since we have a type for the key, we could go the extra mile and make two structs, QueryKey and BorrowedQueryKey, to avoid cloning the suery when looking up the cache(IIRC @abernix  said some queries could be very large).

The idea is that the first struct would hold a String, Option<String> and QueryPlanOptions, and the second one &str, Option<&str> and &QueryPlanOptions. It would also implement Borrow<QueryKey> so it can be used in https://docs.rs/lru/0.7.0/lru/struct.LruCache.html#method.get (there's a similar design in HashMap).

Reply:
In discussion, we agreed to leave this to a follow-up PR. I'll file an issue for this (and other outstanding work) when I finish up this PR.